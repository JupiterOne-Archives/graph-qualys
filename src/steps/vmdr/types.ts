/**
 * Detection target values (`Finding.targets`) pulled from a host asset that
 * serve as additional information for building Finding entities during host
 * detection processing.
 *
 * Some values available on a Qualys `HostAsset` are not available on the
 * `DetectionHost`. The `HostAsset` values are collected in one step and made
 * available to a later step because the Host entities are not added to the
 * `jobState` except as mapped relationship `targetEntity` properties, and are
 * therefore not available for lookup later.
 */
export type HostAssetTargets = {
  fqdn?: string;
  ec2InstanceArn?: string;
  awsAccountId?: string;
  gcpInstanceSelfLink?: string;
  gcpProjectId?: string;
  azureSubscriptionId?: string;
  azureVmSourceId?: string;
};

/**
 * Maps a QWeb Host ID to `HostAssetTargets`.
 */
export type HostAssetTargetsMap = Record<number, HostAssetTargets>;
