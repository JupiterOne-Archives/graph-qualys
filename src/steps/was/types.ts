export type Description = {
  recommendation?: string;
  reference?: string;
  description?: string;
  impact?: string;
};
