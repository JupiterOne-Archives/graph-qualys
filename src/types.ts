/**
 * The integration config managed by the user. This will provide input into the
 * calculation of the `QualysIntegrationConfig` used throughout integration
 * execution.
 */
export type UserIntegrationConfig = {
  qualysUsername: string;
  qualysPassword: string;

  /**
   * The URL of the Qualys API.
   *
   * @example https://qualysapi.qg3.apps.qualys.com
   */
  qualysApiUrl: string;

  /**
   * The minimum number of days since execution time (now) to use when searching
   * for scanned web applications and hosts.
   */
  minScannedSinceDays: string | number;

  /**
   * The minimum number of days since execution time (now) to use when searching
   * for web app findings and host detections.
   */
  minFindingsSinceDays: string | number;

  /**
   * The Qualys vulnerability IDs of detections that should be ingested with
   * results data, used to limit collection of the results property, which can
   * contain an incredible amount of data.
   */
  vmdrFindingResultQids: string | string[];

  /**
   * The severities to use when searching for host detections, used to limit
   * data fetched to only severities a security team wants to ingest.
   */
  vmdrFindingSeverities?: string | string[];

  /**
   * The types of host detections to convert to Findings, used to limit
   * collection to types a security team wants to ingest.
   */
  vmdrFindingTypes?: string | string[];

  ingestWebAppScans: boolean;

  webAppScanApplicationIDFilter: string | string[];

  includeOnlyDetectionTags?: string | string[];
};

export type CalculatedIntegrationConfig = UserIntegrationConfig & {
  /**
   * The start date to use when searching for scanned web applications and
   * hosts, calculated from `minScannedSinceDays`.
   */
  minScannedSinceISODate: string;

  /**
   * The end date to use when searching for scanned hosts, calculated from
   * execution start time.
   */
  maxScannedSinceISODate: string;

  /**
   * The start date to use when searching for web application findings and host
   * detections, calculated from `minFindingsSinceDays`.
   */
  minFindingsSinceISODate: string;

  /**
   * The end date to use when searching for host detections, calculated from
   * execution start time.
   */
  maxFindingsSinceISODate: string;

  /**
   * The severities to use when searching for host detections, used to limit
   * data fetched to only severities a security team wants to ingest. Defaults
   * to `DEFAULT_VMDR_FINDING_SEVERITIES`.
   */
  vmdrFindingSeverityNumbers: number[];

  /**
   * The types of host detections to convert to Findings, used to limit
   * collection to types a security team wants to ingest. Defaults to
   * `DEFAULT_VMDR_FINDING_TYPES`.
   */
  vmdrFindingTypeValues: string[];

  /**
   * The Qualys vulnerability IDs of detections that should be ingested with
   * results data, used to limit collection of the results property, which can
   * contain an incredible amount of data.
   */
  vmdrFindingResultQidNumbers: number[];

  webAppScanApplicationIDs: number[];

  includeOnlyDetectionTags?: string[];
};

/**
 * The integration config expected throughout the integration execution,
 * including all calculated properties.
 */
export type QualysIntegrationConfig = CalculatedIntegrationConfig;

export type PossibleArray<T> = T | T[];

export type Opaque<K, T> = T & { __TYPE__: K };

export type ISODateString = Opaque<'Date', string>;

export type QualyNumericSeverity = 1 | 2 | 3 | 4 | 5 | undefined;

/**
 * Values that conform to the `Finding.numericSeverity` property.
 */
export type NormalizedNumericSeverity = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10;
