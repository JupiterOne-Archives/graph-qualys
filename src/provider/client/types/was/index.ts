export * from './listWebApps';
export * from './listWebAppFindings';
export * from './listWebAppScans';
