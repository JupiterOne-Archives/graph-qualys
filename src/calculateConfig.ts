import snakeCase from 'lodash/snakeCase';

import {
  IntegrationExecutionContext,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';

import {
  DEFAULT_FINDINGS_SINCE_DAYS,
  DEFAULT_SCANNED_SINCE_DAYS,
  DEFAULT_VMDR_FINDING_SEVERITIES,
  DEFAULT_VMDR_FINDING_TYPES,
  DEFAULT_VMDR_FINDING_RESULT_QIDS,
  VALID_VMDR_FINDING_TYPES,
  DEFAULT_WEB_APP_SCAN_APPLICATION_ID_FILTER,
  DEFAULT_INCLUDE_ONLY_DETECTION_TAGS,
} from './constants';
import { CalculatedIntegrationConfig, UserIntegrationConfig } from './types';

export function calculateConfig({
  instance,
  executionHistory,
}: IntegrationExecutionContext<
  UserIntegrationConfig
>): CalculatedIntegrationConfig {
  const config = instance.config;

  const now = Date.now();
  const lastSuccessfulExecutionTime =
    executionHistory.lastSuccessful?.startedOn || 0;

  const minScannedSinceDays = readPropertyAsNumberFromEnvOrConfig(
    config,
    'minScannedSinceDays',
    DEFAULT_SCANNED_SINCE_DAYS,
  );
  const minScannedSinceTime = sinceDaysTime({
    now,
    sinceDays: minScannedSinceDays,
  });
  const minScannedSinceISODate =
    lastSuccessfulExecutionTime > minScannedSinceTime
      ? isoDate(lastSuccessfulExecutionTime)
      : isoDate(minScannedSinceTime);
  const maxScannedSinceISODate = isoDate(executionHistory.current.startedOn);

  const minFindingsSinceDays = readPropertyAsNumberFromEnvOrConfig(
    config,
    'minFindingsSinceDays',
    DEFAULT_FINDINGS_SINCE_DAYS,
  );
  const minFindingsSinceTime = sinceDaysTime({
    now,
    sinceDays: minFindingsSinceDays,
  });
  const minFindingsSinceISODate =
    lastSuccessfulExecutionTime > minFindingsSinceTime
      ? isoDate(lastSuccessfulExecutionTime)
      : isoDate(minFindingsSinceTime);
  const maxFindingsSinceISODate = isoDate(executionHistory.current.startedOn);

  const vmdrFindingSeverityNumbers = readPropertyAsNumberArrayFromEnvOrConfig({
    config,
    propertyName: 'vmdrFindingSeverities',
    defaultValue: DEFAULT_VMDR_FINDING_SEVERITIES,
  });

  const vmdrFindingTypeValues = readPropertyAsStringArrayFromEnvOrConfig({
    config,
    propertyName: 'vmdrFindingTypes',
    defaultValue: DEFAULT_VMDR_FINDING_TYPES,
    validValues: VALID_VMDR_FINDING_TYPES,
  });

  const vmdrFindingResultQidNumbers = readPropertyAsNumberArrayFromEnvOrConfig({
    config,
    propertyName: 'vmdrFindingResultQids',
    defaultValue: DEFAULT_VMDR_FINDING_RESULT_QIDS,
  });

  const webAppScanApplicationIDs = readPropertyAsNumberArrayFromEnvOrConfig({
    config,
    propertyName: 'webAppScanApplicationIDFilter',
    defaultValue: DEFAULT_WEB_APP_SCAN_APPLICATION_ID_FILTER,
  });

  const includeOnlyDetectionTags = readPropertyAsStringArrayFromEnvOrConfig({
    config,
    propertyName: 'includeOnlyDetectionTags',
    defaultValue: DEFAULT_INCLUDE_ONLY_DETECTION_TAGS,
  });

  return {
    ...config,

    minScannedSinceDays,
    minScannedSinceISODate,
    maxScannedSinceISODate,

    minFindingsSinceDays,
    minFindingsSinceISODate,
    maxFindingsSinceISODate,

    vmdrFindingSeverityNumbers,
    vmdrFindingTypeValues,
    vmdrFindingResultQidNumbers,

    webAppScanApplicationIDs,

    includeOnlyDetectionTags,
  };
}

function readPropertyFromEnv(propertyName: string): string | undefined {
  const envName = snakeCase(propertyName).toUpperCase();
  return process.env[envName];
}

function parsePropertyAsNumber(
  propertyName: string,
  value: string | undefined,
): number | undefined {
  if (!value) return undefined;

  const numericValue = Number(value) || undefined;
  if (!numericValue && !!value.trim()) {
    throw new IntegrationValidationError(`Invalid ${propertyName}: ${value}`);
  }

  return numericValue;
}

function parsePropertyAsNumberArray(
  propertyName: string,
  value: string | string[] | undefined,
): number[] | undefined {
  const parseArray = (arr: string[]) => {
    const numbers: number[] = [];
    for (const v of arr) {
      const numericValue = parsePropertyAsNumber(propertyName, v);
      if (numericValue) {
        numbers.push(numericValue);
      }
    }
    if (numbers.length > 0) {
      return numbers;
    }
  };

  if (value && Array.isArray(value)) {
    return parseArray(value);
  } else if (value && value.includes(',')) {
    return parseArray(value.split(','));
  } else {
    const numericValue = parsePropertyAsNumber(propertyName, value);
    if (numericValue) {
      return [numericValue];
    }
  }
}

function parsePropertyAsStringArray(
  propertyName: string,
  value: string | string[] | undefined,
  validValues?: string[],
): string[] | undefined {
  const parseArray = (arr: string[]) => {
    const strings: string[] = [];
    for (const v of arr) {
      strings.push(v.trim());
    }
    if (strings.length > 0) {
      return strings;
    }
  };

  let values: string[] | undefined;

  if (value && Array.isArray(value)) {
    values = parseArray(value);
  } else if (value && value.includes(',')) {
    values = parseArray(value.split(','));
  } else if (value && !/^\s+$/.test(value)) {
    values = [value.trim()];
  }

  if (values && values.length > 0 && validValues && validValues.length > 0) {
    const invalidValues = values.reduce((invalid: string[], e) => {
      if (!validValues.includes(e)) {
        invalid.push(e);
      }
      return invalid;
    }, []);
    if (invalidValues.length > 0) {
      throw new IntegrationValidationError(
        `Invalid ${propertyName}: ${JSON.stringify(invalidValues)}`,
      );
    } else {
      return values;
    }
  } else {
    return values;
  }
}

function readPropertyAsNumberFromEnvOrConfig(
  config: UserIntegrationConfig,
  propertyName: keyof UserIntegrationConfig,
  defaultValue: number,
): number {
  const envValue = readPropertyFromEnv(propertyName);
  const configValue = config[propertyName as string];
  return (
    parsePropertyAsNumber(propertyName, envValue) ||
    parsePropertyAsNumber(propertyName, configValue) ||
    defaultValue
  );
}

function readPropertyAsNumberArrayFromEnvOrConfig({
  config,
  propertyName,
  defaultValue,
}: {
  config: UserIntegrationConfig;
  propertyName: keyof UserIntegrationConfig;
  defaultValue: number[];
}): number[] {
  const envValue = readPropertyFromEnv(propertyName);
  const configValue = config[propertyName as string];
  return (
    parsePropertyAsNumberArray(propertyName, envValue) ||
    parsePropertyAsNumberArray(propertyName, configValue) ||
    defaultValue
  );
}

function readPropertyAsStringArrayFromEnvOrConfig({
  config,
  propertyName,
  defaultValue,
  validValues,
}: {
  config: UserIntegrationConfig;
  propertyName: keyof UserIntegrationConfig;
  validValues?: string[];
  defaultValue: string[];
}): string[] {
  const envValue = readPropertyFromEnv(propertyName);
  const configValue = config[propertyName as string];
  return (
    parsePropertyAsStringArray(propertyName, envValue, validValues) ||
    parsePropertyAsStringArray(propertyName, configValue, validValues) ||
    defaultValue
  );
}

const MILLISECONDS_ONE_DAY = 1000 * 60 * 60 * 24;

function sinceDaysTime(input: { now: number; sinceDays: number }): number {
  return input.now - input.sinceDays * MILLISECONDS_ONE_DAY;
}

function isoDate(time: number) {
  return new Date(time).toISOString().replace(/\.\d{1,3}/, '');
}
