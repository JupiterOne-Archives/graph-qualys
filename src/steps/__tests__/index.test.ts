import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { config } from '../../../test/config';
import { setupQualysRecording } from '../../../test/recording';
import { calculateConfig } from '../../calculateConfig';
import { QualysIntegrationConfig } from '../../types';
import { fetchAccountDetails } from '../account';
import { fetchServices } from '../services';
import {
  fetchScannedHostDetails,
  fetchScannedHostFindings,
  fetchScannedHostIds,
} from '../vmdr';
import { fetchFindingVulnerabilities } from '../vulns';
import {
  fetchAssessments,
  fetchScannedWebAppFindings,
  fetchScannedWebApps,
} from '../was';

jest.setTimeout(1000 * 60 * 1);

let recording: Recording;

afterEach(async () => {
  await recording.stop();
});

test('steps', async () => {
  recording = setupQualysRecording({
    directory: __dirname,
    name: 'steps',
    options: {
      matchRequestsBy: {
        url: {
          query: false,
        },
      },
    },
  });

  const nowTimestamp = 1599865230000; // '2020-09-11T23:00:30Z';
  jest.spyOn(Date, 'now').mockImplementation(() => nowTimestamp);

  const context = createMockStepExecutionContext<QualysIntegrationConfig>({
    instanceConfig: config,
  });

  context.instance.config = calculateConfig(context);

  await fetchAccountDetails(context);
  await fetchServices(context);
  await fetchScannedWebApps(context);
  await fetchScannedWebAppFindings(context);
  await fetchScannedHostIds(context);
  await fetchScannedHostDetails(context);
  await fetchScannedHostFindings(context);
  await fetchFindingVulnerabilities(context);

  // Review snapshot, failure is a regression
  // expect({
  //   numCollectedEntities: context.jobState.collectedEntities.length,
  //   numCollectedRelationships: context.jobState.collectedRelationships.length,
  //   collectedEntities: context.jobState.collectedEntities,
  //   collectedRelationships: context.jobState.collectedRelationships,
  //   encounteredTypes: context.jobState.encounteredTypes,
  // }).toMatchSnapshot();

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Account'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'qualys_account' },
        name: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name'],
    },
  });

  expect(
    context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Service'),
    ),
  ).toMatchGraphObjectSchema({
    _class: ['Service'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: {
          enum: ['qualys_web_app_scanner', 'qualys_vulnerability_manager'],
        },
        name: { type: 'string' },
        version: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['name'],
    },
  });
});

test('fetchScannedHostDetails', async () => {
  recording = setupQualysRecording({
    directory: __dirname,
    name: 'fetchScannedHostDetails',
  });

  const nowTimestamp = 1599865230000; // '2020-09-11T23:00:30Z';
  jest.spyOn(Date, 'now').mockImplementation(() => nowTimestamp);

  const context = createMockStepExecutionContext<QualysIntegrationConfig>({
    instanceConfig: config,
  });

  await fetchAccountDetails(context);
  await fetchServices(context);
  await fetchScannedHostDetails(context);

  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot('fetchScannedHostDetails');
});

test('fetchScannedHostFindings', async () => {
  recording = setupQualysRecording({
    directory: __dirname,
    name: 'fetchScannedHostFindings',
  });

  const nowTimestamp = 1599865230000; // '2020-09-11T23:00:30Z';
  jest.spyOn(Date, 'now').mockImplementation(() => nowTimestamp);

  const context = createMockStepExecutionContext<QualysIntegrationConfig>({
    instanceConfig: config,
  });

  await fetchAccountDetails(context);
  await fetchServices(context);
  await fetchScannedHostFindings(context);

  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot('fetchScannedHostFindings');
});

test('fetchAssessments', async () => {
  recording = setupQualysRecording({
    directory: __dirname,
    name: 'fetchAssessments',
  });

  const nowTimestamp = 1599865230000; // '2020-09-11T23:00:30Z';
  jest.spyOn(Date, 'now').mockImplementation(() => nowTimestamp);

  const context = createMockStepExecutionContext<QualysIntegrationConfig>({
    instanceConfig: config,
  });

  context.instance.config = calculateConfig(context);

  await fetchAccountDetails(context);
  await fetchServices(context);
  await fetchScannedWebApps(context);
  await fetchAssessments(context);

  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot('fetchAssessments');
});
