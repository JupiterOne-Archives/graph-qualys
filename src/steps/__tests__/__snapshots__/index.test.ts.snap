// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAssessments: fetchAssessments 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": undefined,
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": Array [
        "DAST",
      ],
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": undefined,
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": Array [
        "vulnerability-management",
      ],
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": undefined,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 2,
}
`;

exports[`fetchScannedHostDetails: fetchScannedHostDetails 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": Object {
              "Portal-Version": Object {
                "AV2-VERSION": "0.1.0",
                "CA-VERSION": "2.23.0.0",
                "CERTVIEW-VERSION": "2.7.0.0-67",
                "CLOUDVIEW-VERSION": "1.11.0.0-1",
                "CM-VERSION": "1.30.0.0",
                "CS-VERSION": "1.6.8.2-23",
                "FIM-VERSION": "2.3.1.0",
                "IOC-VERSION": "3.0.0.0-5",
                "ITAM-VERSION": "1.2.1.1-1",
                "MDS-VERSION": "2.15.7.0",
                "PM-VERSION": "1.4.3.0-289",
                "PS-VERSION": "1.2.1.0-4",
                "PortalApplication-VERSION": "3.1.2.0-2 OFFICIAL #597 (2020-09-15T15:54:59Z)",
                "QGS-VERSION": "1.2.0.0-6",
                "QUESTIONNAIRE-VERSION": "2.23.1.0",
                "QUESTIONNAIRE__V2-VERSION": "1.10.1.0",
                "THREAT__PROTECT-VERSION": "1.2.0.0",
                "UD-VERSION": "0.0.6",
                "VM-VERSION": "1.0.3",
                "WAF-VERSION": "2.12.4.0",
                "WAS-VERSION": "6.13.1.0",
              },
              "QWeb-Version": Object {
                "SCANNER-VERSION": "12.1.67-1",
                "VULNSIGS-VERSION": "2.5.13-2",
                "WEB-VERSION": "10.3.0.3-2",
              },
            },
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": Array [
        "DAST",
      ],
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": "6.13.1.0",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": Array [
        "vulnerability-management",
      ],
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": "1.0.3",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 2,
}
`;

exports[`fetchScannedHostFindings: fetchScannedHostFindings 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "qualys-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "portal": Object {
              "Portal-Version": Object {
                "AV2-VERSION": "0.1.0",
                "CA-VERSION": "2.23.0.0",
                "CERTVIEW-VERSION": "2.7.0.0-67",
                "CLOUDVIEW-VERSION": "1.11.0.0-1",
                "CM-VERSION": "1.30.0.0",
                "CS-VERSION": "1.6.8.2-23",
                "FIM-VERSION": "2.3.1.0",
                "IOC-VERSION": "3.0.0.0-5",
                "ITAM-VERSION": "1.2.1.1-1",
                "MDS-VERSION": "2.15.7.0",
                "PM-VERSION": "1.4.3.0-289",
                "PS-VERSION": "1.2.1.0-4",
                "PortalApplication-VERSION": "3.1.2.0-2 OFFICIAL #597 (2020-09-15T15:54:59Z)",
                "QGS-VERSION": "1.2.0.0-6",
                "QUESTIONNAIRE-VERSION": "2.23.1.0",
                "QUESTIONNAIRE__V2-VERSION": "1.10.1.0",
                "THREAT__PROTECT-VERSION": "1.2.0.0",
                "UD-VERSION": "0.0.6",
                "VM-VERSION": "1.0.3",
                "WAF-VERSION": "2.12.4.0",
                "WAS-VERSION": "6.13.1.0",
              },
              "QWeb-Version": Object {
                "SCANNER-VERSION": "12.1.67-1",
                "VULNSIGS-VERSION": "2.5.13-2",
                "WEB-VERSION": "10.3.0.3-2",
              },
            },
          },
        },
      ],
      "_type": "qualys_account",
      "createdOn": undefined,
      "displayName": "Qualys - Local Integration",
      "name": "Qualys - Local Integration",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:was",
      "_rawData": Array [],
      "_type": "qualys_web_app_scanner",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Automated Web Application Security Assessment and Reporting",
      "displayName": "Qualys Web Application Scanner",
      "function": Array [
        "DAST",
      ],
      "name": "Qualys Web Application Scanner",
      "updatedOn": undefined,
      "version": "6.13.1.0",
    },
    Object {
      "_class": Array [
        "Service",
      ],
      "_key": "qualys-service:vmdr",
      "_rawData": Array [],
      "_type": "qualys_vulnerability_manager",
      "category": Array [
        "software",
        "other",
      ],
      "createdOn": undefined,
      "description": "Detect, prioritize and remediate vulnerabilities, and monitor using dashboards.",
      "displayName": "Qualys Vulnerability Manager",
      "function": Array [
        "vulnerability-management",
      ],
      "name": "Qualys Vulnerability Manager",
      "updatedOn": undefined,
      "version": "1.0.3",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:was",
      "_toEntityKey": "qualys-service:was",
      "_type": "qualys_account_has_web_app_scanner",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "qualys-account:local-integration-instance",
      "_key": "qualys-account:local-integration-instance|has|qualys-service:vmdr",
      "_toEntityKey": "qualys-service:vmdr",
      "_type": "qualys_account_has_vulnerability_manager",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "qualys_account",
    "qualys_web_app_scanner",
    "qualys_vulnerability_manager",
    "qualys_account_has_web_app_scanner",
    "qualys_account_has_vulnerability_manager",
  ],
  "numCollectedEntities": 3,
  "numCollectedRelationships": 2,
}
`;
